src/gennylib/include/gennylib/PhaseLoop.hpp

```diff
@@ -77,30 +80,69 @@ class IterationCompletionCheck final {
         }
     }
 
-    explicit IterationCompletionCheck(PhaseContext& phaseContext)
-        : IterationCompletionCheck(phaseContext.get<TimeSpec, false>("Duration"),
-                                   phaseContext.get<UIntSpec, false>("Repeat"),
-                                   phaseContext.isNop()) {}
+    explicit IterationChecker(PhaseContext& phaseContext)
+        : IterationChecker(phaseContext.get<TimeSpec, false>("Duration"),
+                           phaseContext.get<IntegerSpec, false>("Repeat"),
+                           phaseContext.isNop()) {
+        auto rateSpec = phaseContext.get<RateSpec, false>("Rate");
+        const auto rateLimiterName =
+            phaseContext.get<std::string, false>("RateLimiterName").value_or("defaultRateLimiter");
+
+        if (rateSpec) {
+            if (!_doesBlock) {
+                throw InvalidConfigurationException(
+                    "Rate must be specified alongside either Duration or Repeat, otherwise there's "
+                    "no guarantee the rate limited operation will run in the correct phase");
+            }
+            _rateLimiter =
+                phaseContext.workload().getRateLimiter(rateLimiterName, rateSpec.value());
+        }
+    }
+
+    bool shouldLimitRate(int64_t currentIteration) const {
+        // Only rate limit if the current iteration is a muliple of the burst size.
+        return _rateLimiter && (currentIteration % _rateLimiter->getBurstSize() == 0);
+    }
 
-    std::chrono::steady_clock::time_point computeReferenceStartingPoint() const {
+    void limitRate(const int64_t currentIteration,
+                   const Orchestrator& o,
+                   const PhaseNumber inPhase) {
+        // This function is called after each iteration, so we never rate limit the
+        // first iteration. This means the number of completed operations is always
+        // `n * GlobalRateLimiter::_burstSize + m` instead of an exact multiple of
+        // _burstSize. `m` here is the number of threads using the rate limiter.
+        if (shouldLimitRate(currentIteration)) {
+            while (true) {
+                auto success = _rateLimiter->consumeIfWithinRate(SteadyClock::now());
+                if (!success && (o.currentPhase() == inPhase)) {
+                    // Add some jitter to avoid threads waking up at once.
+                    std::this_thread::sleep_for(
+                        std::chrono::nanoseconds((_rateLimiter->getRate() + std::rand() % 1000) *
+                                                 _rateLimiter->getNumUsers()));
```

Comment:
plz add comment justifying the multiplication by `getNumUsers`

Reply:
Done, added a comment in getNumUsers() along with explaining what the number is for.